### ISSUE TITLE
Microsoft.Azure.WebJobs.Script.WebHost.DependencyInjection.JobHostServiceProvider.GetRequiredService returns null;

### ISSUE THREAD

**Issue by qsdfplkj** [Azure/azure-functions-host#7539]
Using seviceProvider.GetRequiredService that supports ISupportRequiredService that is Microsoft.Azure.WebJobs.Script.WebHost.DependencyInjection.JobHostServiceProvide returns a null on a interface that is not registered in the serviceProvider.

i'm using 3.1.16 of the microsoft.extensions.dependecyinjection package.

**Comment by v-bbalaiagar** [comment]
Hi @qsdfplkj, Thank you for your feedback! We will check for the possibilities internally and update you with the findings.

**Comment by v-bbalaiagar** [comment]
Hi @qsdfplkj, We would like to have more insights on the scenario to understand it better. Could you please provide us the invocation id of the error, time stamp and region to get the logs and check the cause. Kindly provide the repro steps for us to clone the issue and get the error.

**Comment by qsdfplkj** [comment]
Just create a function app and use the following

    public class INotImplemented
    {
    }

    public class Function1
    {
        public Function1(IServiceProvider serviceProvider)
        {
            // This should throw an error but it doesn't it just returns null;
            INotImplemented required = serviceProvider.GetRequiredService<INotImplemented>();
        }

        [FunctionName("Function1")]
        public void Run([TimerTrigger("*/1 * * * * *")] TimerInfo myTimer, ILogger log)
        {
            log.LogInformation($"C# Timer trigger function executed at: {DateTime.Now}");

        }
    }

**Comment by v-bbalaiagar** [comment]
Hi @qsdfplkj, Thank you for the feedback, we will reproduce this issue and update as appropriate.

**Comment by v-bbalaiagar** [comment]
Hi @qsdfplkj, Apologies for the delayed response, I have tried to reproduce the mentioned scenario. Below are the repro steps:

1.Created function project with Timer trigger function.

![image](https://user-images.githubusercontent.com/74654421/128323051-4579a70e-71e5-409b-baee-06cf5cb335e4.png)

When I debug the app, The function get executed successfully as below and it doesn't returns null.

![image](https://user-images.githubusercontent.com/74654421/128324513-1e0f8e76-41e7-44e5-b4b5-c508f1967c06.png)

Let us know if I am missing a step here.

**Comment by qsdfplkj** [comment]
Yes, locally it runs fine but on azure it uses the JobHostServiceProvider which cause it to return null. this is because the getrequiredservice checks if the actual serviceprovider implements ISupportRequiredService this is supported by the JobHostServiceProvider. And that one just returns null which is  returned by the getrequiredservice.

From: ServiceProviderServiceExtensions

        public static object GetRequiredService(this IServiceProvider provider, Type serviceType)
        {
            if (provider == null)
            {
                throw new ArgumentNullException("provider");
            }

            if (serviceType == null)
            {
                throw new ArgumentNullException("serviceType");
            }

            ISupportRequiredService supportRequiredService = provider as ISupportRequiredService;
            if (supportRequiredService != null)
            {
                return supportRequiredService.GetRequiredService(serviceType); // <------------------------ this returns the null  and skips the exception below. 
            }

            return provider.GetService(serviceType) ?? throw new InvalidOperationException(Resources.FormatNoServiceRegistered(serviceType));
        }


from the JobHostServiceProvider : 

        public object GetRequiredService(Type serviceType)
        {
            return GetService(serviceType, IfUnresolved.Throw);  ///<---- i would assume it throws something but it doesn'???????????????
        }



        private object GetService(Type serviceType, IfUnresolved ifUnresolved)
        {
            if (serviceType == typeof(IServiceProvider))
            {
                return this;
            }

            if (serviceType == typeof(IServiceScopeFactory))
            {
                return this;
            }

            Debug.WriteLine(serviceType.Name);

            return _currentResolver.Container.Resolve(serviceType, ifUnresolved);    //<-------- I assume this just returns a null
        }

The JobHostServiceProvider but that one returns null.
So the testcase would be that JobHostServiceProvider .GetRequiredService(typeof(INotImplemented)) just returns null.
And therefore the ServiceProviderServiceExtensions.GetRequiredService doesn't throw an exception.

You could of course argue that ServiceProviderServiceExtensions.GetRequiredService is wrong. but i think throwing this exception is part of the ISupportRequiredService contract.

**Comment by v-bbalaiagar** [comment]
Thank you for your feedback! We will investigate this further and update as appropriate.

**Comment by v-bbalaiagar** [comment]
Hi @qsdfplkj, Can you please provide the standalone repro that will help us to investigate this issue much easier.

**Comment by qsdfplkj** [comment]
I tried and it wasn’t able. Maybe I could if I would have a nugget package that contains the JobHostServiceProvider somehow I can’t get it from nuget.org. Or Maybe I need to checkout this source. I can try. But I described in detail my issue. What details are you missing??? Maybe I can try later in a few days or so.

**Comment by qsdfplkj** [comment]
For me the following PR will probably fix: https://github.com/dotnet/runtime/pull/58685 (lets see if it gets completed)
But still I think JobHostServiceProvider should take care of it's own.

**Comment by qsdfplkj** [comment]
This is unit test: https://github.com/Azure/azure-functions-host/pull/7664

This fact should be added to JobHostServiceProviderTests.cs

        [Fact]
        public void GetRequiredService_OnJobHost_ShouldThrowAnInvalidOperationExceptionWhenAServiceCannotBeResolved()
        {
            var services = new ServiceCollection();
            var rootScopeFactory = new WebHostServiceProvider(new ServiceCollection());
            var jobHostServiceProvider = new JobHostServiceProvider(services, rootScopeFactory, rootScopeFactory);

            var recordedException = Record.Exception(() => jobHostServiceProvider.GetRequiredService(typeof(IService)));
            Assert.NotNull(recordedException);
            Assert.IsType<InvalidOperationException>(recordedException);
        }

**Comment by liliankasem** [comment]
@qsdfplkj After further investigation this isn't something we're currently able to fix at the moment, it's a potential breaking change for a lot of people. We will pick this up for the next major version release after having a design discussion around this to ensure safe implementation. I'm going to close the corresponding PR (#7664), any future work around this issue should contain a unit test.

**Comment by qsdfplkj** [comment]
Personally I don't see why fixing this is a breaking change. It will show expected behavior. If i call getrequiredservice I expect a non null or exception. 
It will break code that is already working incorrect. But that is what I would be expected. Now I have to debug and discover this issue instead of just getting an exception.

**Comment by liliankasem** [comment]
Yup, I understand that this is expected behaviour. However, breaking code that is already working incorrectly is still breaking code and we need to consider impact on all users before making this change.
